pkg/buses/reconciler.go

```diff
@@ -530,36 +538,15 @@ func (r *Reconciler) syncSubscription(namespace string, name string) error {
 	return nil
 }
 
-func (r *Reconciler) createOrUpdateBus(bus *channelsv1alpha1.Bus) error {
-	if r.bus.GetObjectKind().GroupVersionKind().Kind != bus.Kind ||
-		bus.Namespace != r.bus.GetObjectMeta().GetNamespace() ||
-		bus.Name != r.bus.GetObjectMeta().GetName() {
-		// this is not our bus
-		return nil
-	}
-
-	previousBus := r.bus
-	r.bus = bus.DeepCopy()
-	if !equality.Semantic.DeepEqual(r.bus.GetSpec(), previousBus.GetSpec()) {
-		err := r.handler.onBus(r.bus, r)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (r *Reconciler) createOrUpdateClusterBus(clusterBus *channelsv1alpha1.ClusterBus) error {
-	if r.bus.GetObjectKind().GroupVersionKind().Kind != clusterBus.Kind ||
-		clusterBus.Name != r.bus.GetObjectMeta().GetName() {
-		// this is not our clusterbus
+func (r *Reconciler) createOrUpdateBus(bus channelsv1alpha1.GenericBus) error {
+	if r.ref != NewBusReference(bus) {
+		// not the bus for this reconciler
 		return nil
 	}
 
-	previousBus := r.bus
-	r.bus = clusterBus.DeepCopy()
-	if !equality.Semantic.DeepEqual(r.bus.GetSpec(), previousBus.GetSpec()) {
+	// stash the new bus on the reconciler while retaining the old bus
+	bus, r.bus = r.bus, bus
```

Comment:
r.bus is no longer a DeepCopy of the bus we get back from the busesLister, but during create we make a deepcopy in this same scenario (https://github.com/knative/eventing/pull/420/files#diff-d2cf608fb7c41df730d64ac95f391e7aR313). I don't see why the DeepCopy is needed (am I missing something?) but I imagine these two codepaths should operate the same way.

Reply:
The bus resource on the reconciler should never be mutated.  The general pattern is to create a copy before mutating a resource. The DeepCopy here was an extra layer of protected. With the switch to use GenericBus there is no longer a consistent implementation of DeepCopy to call. 

I tried to add `DeepCopy() GenericBus` to the GenericBus interface, but couldn't get the type checker to accept the generated DeepCopy methods on Bus and ClusterBus since they don't return GenericBus but a struct that implements GenericBus.

To make the behavior more consistent we could call `DeepCopy()` before passing the bus into `createOrUpdateBus()`, or drop the other (hopefully) redundant DeepCopy calls so that's is obvious a copy needs to be created before mutation. Thoughts?

Reply:
Ah, makes sense genericbus would cause issues there. Either way seems reasonable to me, as long as we end up with consistency.

Reply:
Now calling `.DeepCopy()` before passing the bus to `createOrUpdateBus()`.